`modules.tls.validate` is now a thing

waynew opened the pull request:
### What does this PR do?

Add 'validate' to the TLS module, and fixes a couple of bugs

### What issues does this PR fix or reference?

#7424

### Previous Behavior

I still need to double check to see how far back it was, but at some point the `digest` argument was required for a couple of features, but we weren't providing it. I've added those. There were also some encoding (text/unicode vs bytes) problems that popped up when I was running it under Python3, whoops. Running those functions would fail.

### New Behavior

Failing functions no longer fail.

`validate` now exists - one can pass a cert, a CA, and a CRL to confirm that a certificate is valid.

### Tests written?

Yes - including some integration tests to check that we're using pyOpenSSL properly.

### Commits signed with GPG?

Yes

### Things to confirm

Before merging this I do need to check when pyOpenSSL introduced the behavior that has our module broken - I don't want my changes to break anything if that was a reasonably new (<1 year?) change.

Also if there's any code style changes that I should make, please point that out.

Oh, also had a couple of ❓ 
```
# TODO: Should we skip on Windows? Or skip if PyOpenSSL is not installed? -W. Werner, 2019-01-30
# TODO: Should any of these tests be marked destructive? -W. Werner, 2019-01-30
```
I wasn't sure what kind of changes were considered `destructive` - these aren't adding or removing users or anything, but it *is* creating and mucking with a CA. I'm guessing that it's entirely fine though?

waynew committed:
Remove misleading comment.

No clue where it used to point, but there wasn't one of those anywhere
in the codebase... so I removed it.

waynew committed:
Fix digest and type bugs

This is probably not a widely used module, especially for those
installin on Python 3, because there were several errors in accessing
the OpenSSL API, like a lack of `digest` function, and expecting bytes
instead of text.

waynew committed:
Add validate command to tls module

This command validates an x509 certificate against a certificate
authority and an revocation list. Future enhancements might offer the
ability to use a CA file instead.

dwoz reviewed (changes requested):
@waynew Can you look at the linter issues?

waynew committed:
Lint cleanup

waynew committed:
Update CodeClimate settings

https://codeclimate.com/github/saltstack/salt/pull/51461 had a matching
block with mass of 38. Apparently that's too low because that code was
not even the same at all. Bumping it up a bit to avoid false positives.

Also disable argument count checks because we really don't need that
here.

waynew commented:
Okay, I went back and apparently [digest has been around for a very, *very* long time](https://github.com/pyca/pyopenssl/blame/8654f785ccd96a0d6cfb33d521dd891a9178380d/OpenSSL/crypto.py#L1838). Like, it was first introduced in [2014](https://github.com/pyca/pyopenssl/commit/2923dc0f475e0d1a7893c7ca264c95c75f8ba9dd) for CRL. At *least* 5 years ago [the lib expected bytes for the serial](https://github.com/pyca/pyopenssl/blame/4d57590bc7fc93430a1fdacc31bc0cbd9778f678/src/OpenSSL/crypto.py#L1947) and no less than 6 for [the time](https://github.com/pyca/pyopenssl/blame/4d57590bc7fc93430a1fdacc31bc0cbd9778f678/src/OpenSSL/crypto.py#L143).

So, I'm totally comfortable with the fact that there aren't any suitably recent versions of pyOpenSSL that expect anything besides what we're providing here 👍 

@dwoz looks like I fixed the lint issues a while back, too :trollface:
